### PR TITLE
Kaios Push changes : exposed function for register sevice worker and …

### PR DESCRIPTION
**Pull request by mukeshmeena1995**
Kaios Push changes : 
exposed function to register sevice-worker and generate token , uploading same for profile as event request. 
Un-register Token for existing user on onUserLogin. 
Setting Vapid and kaiosPush state on upload-event api response. 
Using Local-Storage for storing VAPID,last-token update ts, kaiospush state. 
On Every App launch it updates token once a day.(as Token can be changed from kaios-servers).

### CODE REVIEW

File: src/CTServiceWorker.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    var globalRedirectPath; // when showing thr url; we need to log to LC before opening up the deep link
+}
+
+
+self.addEventListener('push', function(event) {
+    //self.postMessage()
+   // self.registration.active.postMessage("Hi service worker");
+    // get all the notification data
+    var notificationData = JSON.parse(event.data.text());
+    var title = notificationData['title'];
```
Comment (royson-ct) [inline src/CTServiceWorker.js]: Let's move from bracket to dot notations. Applicable across all files changed.

File: src/CTServiceWorker.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    var notificationOptions = notificationData['notificationOptions'];
+    var data = notificationOptions['data'];
+    var key;
+    if(typeof data !== 'undefined'){
+        key = data['wzrk_id'];
+    }
+    if(typeof key === 'undefined'){
+        key = title;
+    }
+    console.log('Service worker 1 Push event data: ', notificationData);
```
Comment (royson-ct) [inline src/CTServiceWorker.js]: remove `console.log` or update the message. Also we shouldn't print out `notificationData`

File: src/CTServiceWorker.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        //raise notification viewed event
+        fetch(raiseNotificationViewedPath, {'mode': 'no-cors'}); //ignore the response
+    }
+    event.waitUntil(self.registration.showNotification(title, notificationOptions));
+
+});
+
+
+
+self.addEventListener('install', function(event) {
```
Comment (royson-ct) [inline src/CTServiceWorker.js]: remove unwanted code

File: src/CTServiceWorker.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    //     }
+    //     finalDeepLink += '&b=' + encodeURIComponent('button2');
+    // } else {
+        // general click
+        if (typeof notificationData['deepLink'] !== 'undefined') {
+            finalDeepLink += '&r=' + encodeURIComponent(notificationData['deepLink']);
+            silentRequest = false;
+        }
+
+        finalDeepLink += '&b=' + encodeURIComponent('button0');
```
Comment (royson-ct) [inline src/CTServiceWorker.js]: do we have only one button on the notification?

File: src/CTServiceWorker.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    //}
+
+    if (silentRequest) {
+        fireSilentRequest(finalDeepLink);
+    }
+    // else {
```
Comment (royson-ct) [inline src/CTServiceWorker.js]: If it's not a silent request then we do not fire the event? This seems weird 😕

File: src/clevertap.js [hunk 6]
```diff
@@ -46,4 +48,31 @@ export default class CleverTap {
...
+      }
+    } else {
+        // TODO handle anything here ??
+        Utils.log.debug('vapid: ' + Device.getVAPID() + 'notification state:' + Device.getKaiOsNotificationState());
+    }
+  }
+  initiateTokenUpdateIfNeeded() {
+      Utils.log.debug('token updating: for vapid: ' + Device.getVAPID() + 'notification state:' + Device.getKaiOsNotificationState());
+      var lastTokenUpdateTs = Device.getLastTokenUpdateTs(); // when No Registration happens for kaios , it returns current timestamp
+      var oneDay = 24*60*60 ;
```
Comment (royson-ct) [inline src/clevertap.js]: `* 1000` as well since `new Date().getTime()` returns time in ms.

File: src/clevertap.js [hunk 7]
```diff
@@ -46,4 +48,31 @@ export default class CleverTap {
   getAppVersion() {
     return Account.getAppVersion();
   }
+  registerCTNotifications(serviceWorkerPath) {
+    Utils.log.debug('register initiated, vapid: ' + Device.getVAPID());
+    // kaios-Vapid and Push Notification on dashboard should be enabled
+    if(Device.getVAPID() && Device.getKaiOsNotificationState()) {
+      if(this.api !== null) {
```
Comment (royson-ct) [inline src/clevertap.js:66]: I don't think this will ever be false since it's in the constructor.

File: src/device.js [hunk 8]
```diff
@@ -53,4 +53,41 @@ export default class Device {
...
+      } else {
+          StorageManager.save(VAPIDKey, vapid);
+      }
+  }
+    static getVAPID() {
+            const VAPIDKey = StorageManager.getVAPIDKey();
+            var _vapid = StorageManager.read(VAPIDKey);
+            return _vapid;
+    }
+     static setLastTokenUpdateTs(curTs) {
```
Comment (royson-ct) [inline src/device.js:69]: This is not called from anywhere. Am I missing something?

File: src/queueManager.js [hunk 9]
```diff
@@ -220,7 +228,73 @@ export default class QueueManager {
...
+
+                        //Device.setGUID(response.KVAPID);
+                        if (typeof callback === 'function') {
+                            callback(); // call callback to register subscription to kaios servers.
+                        }
+
+                    // handle errors
+                    // request too large
+                } else if (status === 413) {
+                    Utils.log.error('413 on un-register Token request');
```
Comment (royson-ct) [inline src/queueManager.js]: Not a helpful message.

File: src/queueManager.js [hunk 10]
```diff
@@ -220,7 +228,73 @@ export default class QueueManager {
...
+                } else {
+                    Utils.log.error(`kaios vapid request failed with status ${status}.  Will retry.`);
```
Comment (royson-ct) [inline src/queueManager.js]: There is no code to retry here. Update the message or add the code.

**Review by royson-ct — CHANGES_REQUESTED** [review]
Fix linting and console messages. Also cleanup commented code.

File: src/CTServiceWorker.js (hunk 1 above)
Comment (mukeshmeena1995) [reply src/CTServiceWorker.js]: done

File: src/CTServiceWorker.js (hunk 2 above)
Comment (mukeshmeena1995) [reply src/CTServiceWorker.js]: done

File: src/CTServiceWorker.js (hunk 3 above)
Comment (mukeshmeena1995) [reply src/CTServiceWorker.js]: done

File: src/CTServiceWorker.js (hunk 4 above)
Comment (mukeshmeena1995) [reply src/CTServiceWorker.js]: Have removed button for now as no need here.

File: src/clevertap.js (hunk 6 above)
Comment (mukeshmeena1995) [reply src/clevertap.js]: thanks . fixed

File: src/queueManager.js (hunk 9 above)
Comment (mukeshmeena1995) [reply src/queueManager.js]: fixed

File: src/queueManager.js (hunk 10 above)
Comment (mukeshmeena1995) [reply src/queueManager.js]: updated

**Review by royson-ct — APPROVED** [review]
Looks Good 👍

**Review by pwilkniss — APPROVED** [review]

**Pull request merged**